### PR TITLE
dotnet 5.0.205

### PR DESCRIPTION
**Pull request by cho-m**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Experimenting with `dotnet` for Linux...

**EDIT:** Also trying out 5.0.205

### CODE REVIEW

File: Formula/dotnet.rb
Comment (cho-m) [inline Formula/dotnet.rb:36]: Didn't really test LLVM vs. GCC for build time support. Mainly saw build failed on GCC 5 and that Arch & Fedora have `clang`/`llvm` build dependencies, so just switched to LLVM. 

Also, not sure of LLVM as test dependency. Just added since Homebrew test automatically fails with complaint about building `dotnet` inside the `test` block:  
```
==> brew test --verbose dotnet
==> FAILED
==> Testing dotnet
Error: dotnet: failed
An exception occurred within a child process:
  CompilerSelectionError: dotnet cannot be built with any available compilers.
Install Clang or run `brew install gcc`.
```

File: Formula/dotnet.rb
Comment (carlocab) [inline Formula/dotnet.rb:32]: `llvm` also provides a `libunwind`. Since we already have an `llvm` dependency, maybe we can use the one that comes with `llvm`?

Comment (cho-m) [reply Formula/dotnet.rb:32]: Not sure if `llvm` is a runtime dependency though.
Right now, there is no linkage to `llvm`, but there is linkage to `libunwind`.

Used Arch as reference https://archlinux.org/packages/community/x86_64/dotnet-runtime/

Comment (carlocab) [reply Formula/dotnet.rb:32]: Ah yes, that's fine then. `libunwind` is a lighter runtime depdendency anyway.

**Comment by cho-m** [comment]
Should be ready to merge assuming CI passes. Though, `dotnet` occasionally hits CI failures (current CI re-run is due to only Big Sur randomly failing and previous bump also has a random failure).

Also ping @Homebrew/linux since this should provide a Linux bottle to `dotnet` and I think there are some open issues/PRs in `linuxbrew-core`. I removed corresponding label `linux to homebrew-core` as it prevents automerge, which should occur as this is also a bump PR.

File: Formula/dotnet.rb
Comment (danielnachun) [reply Formula/dotnet.rb:36]: Have you tested it with GCC 11?  I think we should usually try that before using LLVM.  If I recall when I tried to fix this in the Linuxbrew repo, it did work with GCC 11 and the blocker was an incompatibility with `icu4c`, which seems to be fixed now.

File: Formula/dotnet.rb
Comment (danielnachun) [reply Formula/dotnet.rb:32]: The `libunwind` library in LLVM doesn't provide all the features of the standalone `libunwind`.  I remember getting very confused when trying to work on `dotnet` before, as I assumed that the LLVM `libunwind` would provide that functionality but I think it was failing.  Ultimately that was why a contributor ended up having to add the standalone `libunwind` as a formula.

**Comment by cho-m** [comment]
CI here is going to be a bit of a pain. Most likely will have to keep rerunning until it passes, which is a bit tedious since it takes 2-3 hours to finish. macOS 10.15 passed yesterday and failed today with same diff.

File: Formula/dotnet.rb
Comment (cho-m) [reply Formula/dotnet.rb:36]: Haven't tried as I was basing on other Linux distros, e.g.
- Fedora: https://src.fedoraproject.org/rpms/dotnet5.0/blob/rawhide/f/dotnet5.0.spec#_81
- Arch: https://github.com/archlinux/svntogit-community/blob/packages/dotnet-core/trunk/PKGBUILD#L23

Also some `dotnet` docs recommend `clang`/`llvm`, e.g.
https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/linux-requirements.md#toolchain-setup
> Install the following packages for the toolchain:
> 
> - cmake
> - llvm-9
> - clang-9

Comment (cho-m) [reply Formula/dotnet.rb:36]: I did see the following comment on https://github.com/dotnet/runtime/pull/53220/files:

> Please note that GNU toolchain support is strictly added for community supported platforms in .NET, where LLVM toolchain is either not available or not as complete. If your platform already has good support for LLVM toolchain, it is recommended (per its official support stauts) to use that instead.

---

So, GCC may work, but the build process may be less maintained as upstream only uses LLVM while GCC is supported by community.

Comment (danielnachun) [reply Formula/dotnet.rb:36]: These are all good points to know.  I think we had problems with linking when using LLVM before, but we fixed those a while back, along with all the other stuff we fixed for LLVM recently.  Since LLVM is only a build/test dependency anyway, it's not creating a huge dependency footprint.  Given these circumstances I think building with LLVM is the right call.

**Comment by danielnachun** [comment]
> Should be ready to merge assuming CI passes. Though, `dotnet` occasionally hits CI failures (current CI re-run is due to only Big Sur randomly failing and previous bump also has a random failure).
> 
> Also ping @Homebrew/linux since this should provide a Linux bottle to `dotnet` and I think there are some open issues/PRs in `linuxbrew-core`. I removed corresponding label `linux to homebrew-core` as it prevents automerge, which should occur as this is also a bump PR.

I'll clean up those issues/PRs in linuxbrew-core once we merge this.  From what I recall, Mono may be the only major language runtime we still need to get bottled on Linux after this.

**Comment by cho-m** [comment]
Will try restarting CI again as it has been queued/stuck on `tap_syntax` for a few hours.

**Comment by cho-m** [comment]
> From what I recall, Mono may be the only major language runtime we still need to get bottled on Linux after this.

For Mono, I wasn't sure if the resources/build-steps need updates as they are from MacSDK packaging steps.

I did try to bottle as is https://github.com/Homebrew/homebrew-core/runs/3175405308?check_suite_focus=true

The current failure is probably due to case-sensitivity. AUR has a patch we can try:
https://aur.archlinux.org/cgit/aur.git/tree/fix_dotnet_case.patch?h=msbuild-git
```diff
diff --git a/mono/build/update_bundled_bits.proj b/mono/build/update_bundled_bits.proj
index eb4cee4d..0f9ebd3a 100644
--- a/mono/build/update_bundled_bits.proj
+++ b/mono/build/update_bundled_bits.proj
@@ -1,5 +1,5 @@
 <Project DefaultTargets="FetchAndUpdateSdksAndNuGets">
-    <Import Project="$(MSBuildThisFileDirectory)\DotnetBitsVersions.props" />
+    <Import Project="$(MSBuildThisFileDirectory)\DotNetBitsVersions.props" />
 
     <Target Name="Build" DependsOnTargets="FetchAndUpdateSdksAndNuGets" />
 ```

**Comment by cho-m** [comment]
And CI 10.15 failed again. Maybe a race condition or parallelization issue. Will retry again.
```
  Build FAILED.
  
  /private/tmp/dotnet-20210727-43148-11wdyt9/artifacts/src/msbuild.5e4b48a27efce55a613664b58d353ab4c8d1f6c1/src/Tasks/system.design/system.design.txt : error MSB3103: Invalid Resx file. The process cannot access the file '/private/tmp/dotnet-20210727-43148-11wdyt9/artifacts/src/msbuild.5e4b48a27efce55a613664b58d353ab4c8d1f6c1/src/Tasks/system.design/system.design.txt' because it is being used by another process. [/private/tmp/dotnet-20210727-43148-11wdyt9/artifacts/src/msbuild.5e4b48a27efce55a613664b58d353ab4c8d1f6c1/src/Tasks/Microsoft.Build.Tasks.csproj]
      0 Warning(s)
      1 Error(s)
```

**Comment by cho-m** [comment]
CI finally passed 🎉

**Review by danielnachun — APPROVED** [review]

**Pull request closed without merging**
